Table Mutli-selection documentation (#1874)
hashicorp/design-system: website/docs/components/table/partials/guidelines/guidelines.md

- We recommend using `medium` cell density by default.
- If content is complex or a smaller data set (e.g., a Table of basic user data), `tall` cell density allows for more breathing room surrounding the content.
- If content is largely string/text-based, `short` allows for more content to be displayed within the page.
- While denser content allows for more rows to be displayed within a single page, it also makes comprehension and scanning more difficult.

## Multi-Select

> Multi-select enables users to select multiple rows from a dataset, significantly enhancing the efficiency of batch operations like delete, move, or update. This functionality is important for effective data management, especially in large datasets.

Reviewer: Maybe a complex answer here, but I think we need to differentiate between rows, tables, results, and datasets. These have different contextual meanings that relate to each other, for example:

- A data set consists of many results
- A table consists of many rows
- A data set can be rendered within a table, the table is the UI representation of the data set. but it can also be rendered in other ways.

So, in this scenario I would maybe approach from this angle:

- "Multi-select enables users to select multiple rows in a table…" or
- "Multi-select enables users to select multiple results from a data set when rendered as a table…"

Not perfect, but I think there's a benefit in the clarity of terminology here.

Reviewer: Also, global note: in the filtering docs I intentionally used "data set" rather than "dataset". We should strive for consistency here.

Reviewer: To summarize, maybe consider this:

> Multi-select enables users to select multiple rows from a table, allowing for batch actions like delete or update.

The point about effective data management seems a bit broad/ambiguous, is there a way to more accurately express the end-user value?

> Multi-select enables users to select multiple rows from a dataset, significantly enhancing the efficiency of batch operations like delete, move, or update. This functionality is important for effective data management, especially in large datasets.

Author: ```suggestion
Multi-select enables users to select multiple rows from a table. Users benefit from multi-select by experiencing enhanced efficiency and reduced errors, improving overall data management. This feature lets users quickly perform bulk actions on multiple rows, significantly increasing productivity.

```

!!! Info 

> Avoid using indeterminate states in individual rows. It's more effective at the global selection checkbox to reduce confusion.

Reviewer: I think you can be more pointed here and essentially say that indeterminate state isn't supported in individual rows. Think back to what we discussed about parent/child relationships here. The rows (in the majority of scenarios) don't have the concept of children, therefore they can only be considered on/off, checked/un-checked.

> Avoid using indeterminate states in individual rows. It's more effective at the global selection checkbox to reduce confusion.

Author: ```suggestion
While you’re able to have the indeterminate state in the Cell / Select component, in the ember component, it is not achievable. 
```

!!!

<video width="100%" controls loop>
  <source
    src="/assets/components/table/multi-selection-interaction-01.mp4"
    type="video/mp4"
  />
</video>

> ### Suggested Application

Author: ```suggestion
### Intended interaction
```

> - For global selection and actions we recommend a master checkbox at the top of the table to select all dataset items in one click, simplifying bulk operations. The interface should allow operations like select all (rows) or deselct all (rows), and delete, or move to be applied to all selected rows for quick and efficient control across multiple pages.

Reviewer: Couple of points here:

- This can be much more pointed and direct as well, shouldn't be worded as a recommendation because of how this feature works in the Ember component. The property that determines selection is `isSelectable`, if it's set to `{{true}}` at the top level of the table, the relevant UI elements will be rendered by default.
- This bullet point combines information about the intended UI, functionality, and how to handle bulk actions, which is pretty confusing. I think this should be decoupled a bit and each item addressed individually.

> ### Suggested Application
> 
> - For global selection and actions we recommend a master checkbox at the top of the table to select all dataset items in one click, simplifying bulk operations. The interface should allow operations like select all (rows) or deselct all (rows), and delete, or move to be applied to all selected rows for quick and efficient control across multiple pages.

Author: ```suggestion
### Suggested Application

Header Column /  Select is the master checkbox that renders a checked state or indeterminate state when Cell / Selection is selected. 
Cell / Selection is the component that enables multi-selection. It is always in the leading position of the row and can be rendered with {{isStriped}} or Medium, Short, and Tall densities

```

> - It is beneficial to display a dynamically updating count of selected rows. This feature provides a continuous and straightforward overview of user selections, adjusting in real-time as selections are made or changed.
> - When pagination is used in a table, maintain user selections as consistent and manageable across different pages.

Author: ```suggestion
Selecting a single row highlights it for actions like edit or delete.
Selecting the top-level header selects all rows on the current page.
Selecting the top-level header automatically selects all visible rows on the current page, including any rows previously chosen individually.
The 'Select all' usually affects only the current page in paginated tables. While in comparison, a global selection will select all rows on all pages (i.e., select all 53 users)
```

- We recommend using `medium` cell density by default.
- If content is complex or a smaller data set (e.g., a Table of basic user data), `tall` cell density allows for more breathing room surrounding the content.
- If content is largely string/text-based, `short` allows for more content to be displayed within the page.
- While denser content allows for more rows to be displayed within a single page, it also makes comprehension and scanning more difficult.

## Multi-Selection

> Multi-selection enables users to select multiple rows in a table, enhancing efficiency and minimizing errors. This feature facilitates bulk operations, boosting productivity and optimizing data management.

Reviewer: ```suggestion
Multi-selection enables users to select multiple rows in a table, facilitating bulk operations, actions, and editing.
```

I still think this point around productivity and data management is a bit irrelevant (at least for this type of documentation). Though curious what other folks thing.

> To enable multi-selection in a table:

Reviewer: I would pivot this to be more like:

> Multi-selection consists of:

Then list each component out.

1. `Header Column / Selection` Component:

> Position it in the table's header row. This acts as a master control, enabling you to either select all rows simultaneously or deselect them.

Reviewer: ```suggestion
Used in the header row of the Table. This acts as a master control, enabling end-users to either select or deselect all rows simultaneously.
```

![Example of centered content within a table](/assets/components/table/multi-select-header.png)

2. `Cell / Selection` Component:

> Integrate this into each table row. It allows for individual row selection and aligns with the Header Column component, reflecting the overall selection status of the table.

Reviewer: ```suggestion
Used in the start position each Table row. It allows for individual row selection and aligns with the Header Column component, reflecting the overall selection status of the table.
```

![Example of centered content within a table](/assets/components/table/multi-select-cells.png)

!!! Info 

> In the Ember component, unlike the `cell/select` figma component, achieving an indeterminate state is not possible. 

Reviewer: Maybe consider something like this:

> While technically possible within the Figma component, the Checkbox within selected rows should not reflect an indeterminate state.

You could also include something about the child/parent relationship between the Header selection and the Cell selection.

Reviewer 2: @andgen404 I don't quite understand what you mean here. Are you talking about having an indeterminate state on the checkboxes in the table body rows (vs. in the table header)? Because that doesn't make sense to do anyway. I would be very confused by this as a consumer.

> In the Ember component, unlike the `cell/select` figma component, achieving an indeterminate state is not possible. 

Reviewer 2: ```suggestion
In the Ember component, unlike the `cell/select` Figma component, achieving an indeterminate state is not possible. 
```

!!!

<video width="100%" controls loop>
  <source
    src="/assets/components/table/multi-selection-interaction-01.mp4"
    type="video/mp4"
  />
</video>

> ### Intended interaction

Reviewer: I think this section might need a bit more fleshing out. Maybe consider breaking it down into a couple of different sections, could look something like this:

- Intended interaction
    - Within a simple table (this is where you can discuss what will happen with the selection states if the table is a single page/doesn't use pagination).
    - Within a paginated table
    - (maybe something else here?)

> - Selecting an individual row highlights it, indicating it is ready for actions such as editing or deletion.

Reviewer: ```suggestion
- Selecting an individual row indicates it is ready for actions such as editing or deletion.
```

Using "highlights" is confusing (and probably loaded) term here.

> - Activating the top-level header checkbox selects every row on the current page. 

Reviewer: ```suggestion
- Interacting with the top-level header checkbox selects every row on the current page. 
```

Could also say "Checking" as that refers to the actual property.

Reviewer 2: I think "checking" would be much clearer.

>      - This action ensures that all visible rows on the page, including those previously selected individually, are chosen collectively.

Reviewer: This reads a bit weird. Maybe consider discussing when the `Header Column / Selection` will:

- render as non-checked
- render as indeterminate
- render as checked

And how the selected rows are impacted by each one of these states.

Reviewer 2: Agree that this wording is confusing. Just say what it actually does, ie, it selects all visible rows on the page.

> - The 'Select all' function impacts only the rows displayed on the current page in paginated tables. In contrast, a global 'Select all' option encompasses every row across all pages (for example, selecting all 53 users in a dataset).

Reviewer: I wouldn't use the term "Select all" here because 1) there isn't a label present that expresses this concept (even if we mean to be coupled with the Header Column / Selection), and 2) because this implies that we also provide a "Select all" function with a paginated table, which we do not.

> - The 'Select all' function impacts only the rows displayed on the current page in paginated tables. In contrast, a global 'Select all' option encompasses every row across all pages (for example, selecting all 53 users in a dataset).

Reviewer 2: ```suggestion
- The “Select all” function impacts only the rows displayed on the current page in paginated tables. In contrast, a global “Select all” option encompasses every row across all pages (for example, selecting all 53 users in a dataset).
```

- We recommend using `medium` cell density by default.
- If content is complex or a smaller data set (e.g., a Table of basic user data), `tall` cell density allows for more breathing room surrounding the content.
- If content is largely string/text-based, `short` allows for more content to be displayed within the page.
- While denser content allows for more rows to be displayed within a single page, it also makes comprehension and scanning more difficult.

## Multi-Selection

Multi-selection enables users to select multiple rows in a table, facilitating bulk operations, actions, and editing.

Multi-selection consist of: 

1. `Header Column / Selection` Component:
Used in the header row of the Table. This acts as a master control, enabling end-users to either select or deselect all rows simultaneously.

> 

Author: ```suggestion
```

Used in the start position each Table row. It allows for individual row selection and aligns with the Header Column component, reflecting the overall selection status of the table.

![Example of centered content within a table](/assets/components/table/multi-select-cells.png)

!!! Info 

In the Ember component, unlike the `cell/select` Figma component, achieving an indeterminate state is not possible. 
!!!

### Intended interaction

#### Within a Simple Table (Single Page):

- Selecting an Individual Row: Indicates that the row is ready for actions such as editing or deletion.
- Interaction with the Top-Level Header Checkbox: Non-Checked State: Initially, the checkbox is unmarked when none of the rows are selected.
- Checkbox Activation: Clicking the checkbox selects every row in the table, including any previously chosen individually, ensuring collective selection.
- Indeterminate State: The checkbox shows an indeterminate state when some but not all rows are selected, indicating a partial selection.

<video width="100%" controls loop>
  <source
    src="/assets/components/table/multi-selection-interaction-01.mp4"
    type="video/mp4"
  />
</video>

> #### Within a paginated table: 
> 
> - Global Actions:
> In addition to page-specific actions, paginated tables may benefit from global actions like a "Select all users" feature. This would apply to all rows across every page in the table, enabling actions on a larger scale (e.g., selecting all 53 users in a dataset or deselecting). 
> 
> <video width="100%" controls loop>
>   <source
>     src="/assets/components/table/multi-selection-interaction-02.mp4"
>     type="video/mp4"
>   />
> </video>

Author: ```suggestion
<!--TODO: Add in for future use of pattern documentation #### Within a paginated table: 

- Global Actions:
In addition to page-specific actions, paginated tables may benefit from global actions like a "Select all users" feature. This would apply to all rows across every page in the table, enabling actions on a larger scale (e.g., selecting all 53 users in a dataset or deselecting). 

<video width="100%" controls loop>
  <source
    src="/assets/components/table/multi-selection-interaction-02.mp4"
    type="video/mp4"
  />
</video>-->
```
